<a class="anchor" id="0"></a>
# **AdaBoost Classifier**

# **AdaBoost implementation in Python** <a class="anchor" id="6"></a>

### Import libraries <a class="anchor" id="6.1"></a>

In [36]:
import pandas as pd
import numpy as np

### Load dataset <a class="anchor" id="6.2"></a>

In [37]:
df = pd.read_csv('./data/Iris.csv')


### EDA <a class="anchor" id="6.3"></a>

### Preview dataset

In [38]:
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

### View summary of dataframe

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


### Declare feature vector and target variable

In [40]:
X = df[['SepalLengthCm','SepalWidthCm','PetalLengthCm','PetalWidthCm']]
X.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0            5.1           3.5            1.4           0.2
1            4.9           3.0            1.4           0.2
2            4.7           3.2            1.3           0.2
3            4.6           3.1            1.5           0.2
4            5.0           3.6            1.4           0.2

In [41]:
y = df['Species']
y.head()

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: Species, dtype: object

In [42]:
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()
y=le.fit_transform(y)

###  Split dataset into training set and test set <a class="anchor" id="6.4"></a>

In [43]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

###  AdaBoost model with Decision Tree Classifier as estimator <a class="anchor" id="6.5"></a>

In [44]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

dtc = DecisionTreeClassifier()
abc = AdaBoostClassifier(n_estimators=50, learning_rate=1, random_state=0, estimator=dtc)
model1 = abc.fit(X_train, y_train)
y_pred = model1.predict(X_test)
print("AdaBoost Classifier Model Accuracy:", accuracy_score(y_test, y_pred))

AdaBoost Classifier Model Accuracy: 0.9333333333333333


/home/ayush/Desktop/university/ml/ise1/code/venv/lib/python3.11/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


#### AdaBoost with DecisionTree as Estimator is overfitting, which is a problem, we can try different estimators to get a suitable result

###  AdaBoost model with Support Vector Classifier as estimator  <a class="anchor"></a>

In [45]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

svc=SVC(probability=True, kernel='linear')
abc =AdaBoostClassifier(n_estimators=50, learning_rate=1, random_state=0, estimator=svc)
model2 = abc.fit(X_train, y_train)
y_pred = model2.predict(X_test)
print("Model Accuracy with SVC Base Estimator:",accuracy_score(y_test, y_pred))


/home/ayush/Desktop/university/ml/ise1/code/venv/lib/python3.11/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


Model Accuracy with SVC Base Estimator: 0.9111111111111111


- In this case, we have got a classification rate of 91.11%, which is considered as a very good accuracy.


- In this case, SVC Base Estimator is getting realistic accuracy then Decision tree estimator.